src/o365/spfx/commands/project/project-externalize/rules/PnPJsRule.ts

```diff
@@ -40,20 +45,52 @@ export class PnPJsRule extends BasicDependencyRule {
     }
   ];
 
-  public visit(project: Project): Promise<ExternalizeEntry[]> {
+  public visit(project: Project): Promise<[ExternalizeEntry[],FileEditSuggestion[]]> {
     const findings = this.pnpModules
       .map(x => this.getModuleAndParents(project, x.key))
       .reduce((x, y) => [...x, ...y]);
-
-    if (findings.filter(x => x.key !== '@pnp/pnpjs').length > 0) {
+    const files = this.getEntryFilesList(project);
+    const rawfileEdits = this.pnpModules.filter(x => findings.find(y => y.key === x.key) !== undefined)
+        .filter(x => x.shadowRequire !== undefined)
+        .map(x => files.map(y => ({
+          action: "add",
+          path: y,
+          targetValue: x.shadowRequire
+        } as FileEditSuggestion)))
+    const fileEdits = rawfileEdits.length > 0 ? rawfileEdits.reduce((x, y) => [...x, ...y]) : [];
+    if (findings.filter(x => x.key && x.key !== '@pnp/pnpjs').length > 0) {
       findings.push({
         key: 'tslib',
         globalName: 'tslib',
         path: `https://unpkg.com/tslib@^1.10.0/tslib.js`
       });
+      fileEdits.push(...files.map(x => ({
+          action: "add",
+          path: x,
+          targetValue: 'require(\"tslib\");'
+      } as FileEditSuggestion)));
     }
-
-    return Promise.resolve(findings);
+    
+    return Promise.resolve([findings, fileEdits]);
+  }
+  private getEntryFilesList(project: Project): string[] {
+    const result = [...this.getComponents(project, 'webparts', ['WebPart']), ...this.getComponents(project, 'extensions', ['ApplicationCustomizer', 'CommandSet'])];
+    return result;
+  }
+  private getComponents(project:Project, componentPathSegment: string, appendixes: string[]): string[] {
```

Comment:
We should not be walking through files in rules. Instead, if we need access to specific files, we should extend the project model and its collection to retrieve everything that's necessary there and avoid that we're traversing the same files multiple times across the different rules.

Reply:
the files I need here are very specific to this rule and I didn't want to impact the whole project structure, which is used by other commands, just for this one rule. I'd advise against bubbling the very specific logic to the project.

Reply:
If you look at the `Project` model, you can see that it has a designated collection for `.ts` files and if you look at the `getProject` method in the `project upgrade` command, you see how we discover the different code and manifest files. I'd be for using a consistent approach across the commands to simplify long-term maintenance. Now that we have multiple commands working on projects, perhaps it would even make sense to centralize retrieving project to drive consistency and avoid code duplication.

Reply:
I missed the fact that getProject from upgrade had the manifests, much simpler to rely on that rather than writing my own code!
And yes upgrade and externalize could have a lot of code in common, and it'd probably be a good thing to refactor, that was one of my original comments in the first externalize PR.
I took a shot at it with the last commit I just pushed, then we can gradually move things over time if the approach is good.
Thanks for the constructive comment!